### PR TITLE
feat: add ability to follow links in img and source elements

### LINKED ISSUE #1103: Exported server routes returning binary files get corrupted
**Describe the bug**
When a route points to a server route that emits a binary file (pdf, image etc.) and the site is `export`ed, the files on disk get corrupted because export only supports text results from server routes.

**To Reproduce**
- Create a server route that returns a binary file (image, PDF)
- Link to the server route inside a Svelte component
- Export the site
- Find the exported file of the server route on disk

Result: the file is corrupt and cannot be opened.

**Expected behavior**
The result of server routes emitting binary data should be properly stored on disk when exporting.

**Information about your Sapper Installation:**
- Sapper master branch (v0.27.10)
- Exported site
- Checked with rollup, but should be same for webpack

**Severity**
Blocking my attempt to load all data including binary assets (documents, images) from a headless CMS.

### LINKED ISSUE #1104: Sapper does not crawl img and source elements
**Is your feature request related to a problem? Please describe.**
Sapper only follows links in `<a>` elements. 

**Describe the solution you'd like**
Sapper should crawl links in the `src` and `srcset` attributes in `` and `<source>` tags and export them to disk if they refer to server routes.

**Describe alternatives you've considered**
Manually copying the related files to the static directory, but that is not feasible when using Sapper for a headless CMS.

**How important is this feature to you?**
I'm developing a prototype frontend for a headless CMS and this is a huge problem. My server routes load the images from the CMS and transform them using [sharp](https://www.npmjs.com/package/sharp) (resizing, type conversion etc for responsive images). Without this feature I cannot use the automatic conversion and implement an unrelated build step for this.

**Additional context**
PR #938 already proses a partial solution but I have a PR in the available that adds support for srcset attributes.

### PR DESCRIPTION
**Pull request by aloker**
This PR supersedes #938. It implements the following new behavior:

- In addition to a-elements, `export` now also follows links found in ``, ` and `<source srcset>` and exports them if they point to server routes. 

- As a side effect, links to other server routes emitting binary files also properly export the binary files. Before, binary files were not correctly exported.

Fixes #1103
Fixes #1104

### CODE REVIEW

File: src/api/export.ts [hunk 1]
```diff
@@ -115,19 +130,26 @@ async function _export({
 			if (pathname !== '/service-worker-index.html') {
 				file = file === '' ? 'index.html' : `${file}/index.html`;
 			}
-			body = minify_html(body);
+
+			if(typeof(body)!=='string'){
```
Comment (benmccann) [inline src/api/export.ts:134]: nit: should add a spaces and I don't think you need the parens:

    if (typeof body !== 'string') {

File: src/api/export.ts [hunk 2]
```diff
@@ -176,31 +200,47 @@ async function _export({
...
-					const base_href = base_match && get_href(base_match[1]);
-					const base = resolve(url.href, base_href);
-
-					let match;
-					let pattern = /<a ([\s\S]+?)>/gm;
-
-					while (match = pattern.exec(cleaned)) {
-						const attrs = match[1];
-						const href = get_href(attrs);
+					if(typeof(body) === 'string') {
```
Comment (benmccann) [inline src/api/export.ts:210]: you could combine this with the previous `if` check to remove a level of indentation, which would make things easier to read:

    if (pathname !== '/service-worker-index.html' && typeofbody === 'string') {

File: src/api/export.ts [hunk 3]
```diff
@@ -176,31 +200,47 @@ async function _export({
...
+									break;
+							}
 
-						if (href) {
-							const url = resolve(base.href, href);
+							hrefs = hrefs.filter(Boolean);
 
-							if (url.protocol === protocol && url.host === host) {
-								tasks.push(handle(url));
+							for(const href of hrefs) {
```
Comment (benmccann) [inline src/api/export.ts:239]: nit: space after `for`

File: src/api/export.ts [hunk 4]
```diff
@@ -176,31 +200,47 @@ async function _export({
...
+						const base = resolve(url.href, base_href);
+
+						let match;
+						let pattern = /<(a|img|source)\s+([\s\S]+?)>/gm;
+
+						while (match = pattern.exec(cleaned)) {
+							let hrefs:string[] = [];
+							const element = match[1];
+							const attrs = match[2];
+							switch(element){
```
Comment (benmccann) [inline src/api/export.ts:224]: nit: space after `switch` and before `{`

File: src/api/export.ts [hunk 5]
```diff
@@ -246,3 +286,29 @@ function get_href(attrs: string) {
...
+	const match = /src\s*=\s*(?:"(.*?)"|'(.*?)'|([^\s>]*))/.exec(attrs);
+	return match && (match[1] || match[2] || match[3]);
+}
+
+
+export function get_srcset_urls(attrs: string) {
+	const results:string[] = []
+	// Note that the srcset allows any ASCII whitespace, including newlines.
+	const match = /srcset\s*=\s*(?:"(.*?)"|'(.*?)'|([^\s>]*))/s.exec(attrs);
+	if(match){
```
Comment (benmccann) [inline src/api/export.ts:300]: I would invert this check to reduce indentation:

```
if (!match) {
  return results;
}
```

File: src/api/export.ts [hunk 6]
```diff
@@ -246,3 +286,29 @@ function get_href(attrs: string) {
...
+		const attrContent = match[1] || match[2] || match[3];
+		// Parses the content of the srcset attribute.
+		// The regexp is modelled after the srcset specs (https://html.spec.whatwg.org/multipage/images.html#srcset-attribute)
+		// and should cover most reasonable cases.
+		const regex = /\s*([^\s,]\S+[^\s,])\s*((?:\d+w)|(?:-?\d+(?:\.\d+)?(?:[eE]-?\d+)?x))?/gm;
+		let subMatches = regex.exec(attrContent)
+		while(subMatches){
```
Comment (benmccann) [inline src/api/export.ts:307]: nit: space after `while`

File: test/apps/export/src/routes/img/[slug].png.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+const path = require('path');
+const cwd = process.cwd();
+
+export function get(req, res, next) {
+
+	const { slug } = req.params;
+	const image = path.join(cwd, `/content/${slug}.png`);
+
+	let s = fs.createReadStream(image);
+	s.on('open', function () {
```
Comment (benmccann) [inline test/apps/export/src/routes/img/[slug].png.js:11]: nit: remove space `function() {`

File: test/apps/export/src/routes/pdfs/[slug].pdf.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+const path = require('path');
+const cwd = process.cwd();
+
+export function get(req, res, next) {
+
+	const { slug } = req.params;
+	const image = path.join(cwd, `/content/${slug}.pdf`);
+
+	let s = fs.createReadStream(image);
+	s.on('open', function () {
```
Comment (benmccann) [inline test/apps/export/src/routes/pdfs/[slug].pdf.js:11]: nit: remove space `function() {`

File: test/apps/export/test.ts [hunk 9]
```diff
@@ -49,6 +52,20 @@ describe('export', function() {
 		const input = readFileSync(`${__dirname}/static/test.pdf`)
 		const output = readFileSync(`${__dirname}/__sapper__/export/test.pdf`)
 		assert.ok(input.equals(output))
+	});
+
+	it('does not corrupt image files from server routes', () => {
+		for(const file of ['example-192.png', 'example-512.png']) {
```
Comment (benmccann) [inline test/apps/export/test.ts:58]: nit: add space after `for`

**Comment by Conduitry** [comment]
Closing in favor of #1398, which is built on this. I've included you as a co-authored-by on the commit. Thank you!

**Pull request closed without merging**